File: src/data/memory.ts
import { Database } from 'bun:sqlite';
import { homedir } from 'os';
import { join, basename } from 'path';
import { existsSync } from 'fs';
import type { OcMemoryChunk } from '../types.js';

const DB_PATH = join(homedir(), '.openclaw', 'memory', 'main.sqlite');

interface ChunkRow {
  id: string;
  path: string;
  source: string;
  start_line: number;
  end_line: number;
  model: string;
  text: string;
  updated_at: number;
}

export function loadMemory(): OcMemoryChunk[] {
  if (!existsSync(DB_PATH)) return [];
  try {
    const db = new Database(DB_PATH, { readonly: true });
    const rows = db.query<ChunkRow, []>(
      'SELECT id, path, source, start_line, end_line, model, text, updated_at FROM chunks ORDER BY updated_at DESC'
    ).all();
    db.close();
    return rows.map(r => ({
      kind: 'memory' as const,
      id: r.id,
      name: `${basename(r.path)}:${r.start_line}-${r.end_line}`,
      path: r.path,
      source: r.source,
      startLine: r.start_line,
      endLine: r.end_line,
      model: r.model,
      text: r.text,
      updatedAt: r.updated_at,
    }));
  } catch {
    return [];
  }
}

File: internal/data/memory.go
package data

import (
	"database/sql"
	"fmt"
	"os"
	"path/filepath"

	_ "modernc.org/sqlite"
)

func LoadMemory() []OcMemoryChunk {
	home, _ := os.UserHomeDir()
	dbPath := filepath.Join(home, ".openclaw", "memory", "main.sqlite")

	if _, err := os.Stat(dbPath); err != nil {
		return nil
	}

	db, err := sql.Open("sqlite", "file:"+dbPath+"?mode=ro")
	if err != nil {
		return nil
	}
	defer db.Close()

	rows, err := db.Query(
		`SELECT id, path, source, start_line, end_line, model, text, updated_at
		 FROM chunks ORDER BY updated_at DESC`,
	)
	if err != nil {
		return nil
	}
	defer rows.Close()

	var items []OcMemoryChunk
	for rows.Next() {
		var r struct {
			id        string
			path      string
			source    string
			startLine int
			endLine   int
			model     string
			text      string
			updatedAt int64
		}
		if err := rows.Scan(&r.id, &r.path, &r.source, &r.startLine, &r.endLine, &r.model, &r.text, &r.updatedAt); err != nil {
			continue
		}
		name := fmt.Sprintf("%s:%d-%d", filepath.Base(r.path), r.startLine, r.endLine)
		items = append(items, OcMemoryChunk{
			ID:        r.id,
			Name_:     name,
			Path:      r.path,
			Source:    r.source,
			StartLine: r.startLine,
			EndLine:   r.endLine,
			Model:     r.model,
			Text:      r.text,
			UpdatedAt: r.updatedAt,
		})
	}
	return items
}
